lib/TokenManager.js

```diff
@@ -120,25 +129,31 @@ function refresh(sdk, tokenMgmtRef, storage, key) {
   }
 
   // Remove existing autoRefresh timeout for this key
-  clearRefreshTimeout(tokenMgmtRef, key);
-
-  return sdk.token.refresh(token)
-  .then(function(freshToken) {
-    add(sdk, tokenMgmtRef, storage, key, freshToken);
-    tokenMgmtRef.emitter.emit('refreshed', key, freshToken, token);
-    return freshToken;
-  })
-  .fail(function(err) {
-    if (err.name === 'OAuthError') {
-      remove(tokenMgmtRef, storage, key);
-      emitExpired(tokenMgmtRef, key, token);
-    }
-    throw err;
-  });
+  clearExpireEventTimeout(tokenMgmtRef, key);
+
+  // Store the refresh promise state, to avoid refreshing again
+  if (!tokenMgmtRef.refreshPromise[key]) {
+    tokenMgmtRef.refreshPromise[key] = sdk.token.refresh(token)
+    .then(function(freshToken) {
+      add(tokenMgmtRef, storage, key, freshToken);
+      tokenMgmtRef.emitter.emit('refreshed', key, freshToken, token);
+
+      return freshToken;
+    })
+    .fail(function(err) {
+      if (err.name === 'OAuthError') {
+        remove(tokenMgmtRef, storage, key);
+        emitError(tokenMgmtRef, err);
+      }
+
+      throw err;
+    });
+  }
+  return tokenMgmtRef.refreshPromise[key];
 }
 
 function clear(tokenMgmtRef, storage) {
-  clearRefreshTimeoutAll(tokenMgmtRef);
```

Comment:
What about `refreshPromise`? Also, a bit of an edge case - but if they choose to `clear` when we have a pending refresh action, what happens?

Reply:
Its likely that the token would get refreshed and stored shortly after.

To account for this scenario, I'll push a change where we're clearing the refresh keys when `clearRefreshTimeoutAll` is called - i.e. (`tokenManager.clear()`) is called externally.

Reply:
That doesn't solve the problem, does it?